ckanext/emailasusername/templates/user/new_user_form.html

```diff
@@ -2,8 +2,35 @@
 
 <form id="user-register-form" action="" method="post">
   {{ form.errors(error_summary) }}
-  {{ form.input("name", id="field-username", label=_("Username"), placeholder=_("username"), value=data.name, error=errors.name, classes=["control-medium"], is_required=True) }}
-  {{ form.input("fullname", id="field-fullname", label=_("Full Name"), placeholder=_("Joe Bloggs"), value=data.fullname, error=errors.fullname, classes=["control-medium"]) }}
+
+  {% if h.get_auto_generate_username_from_fullname() %}
+    <div id="username-input-field">
+      {{ form.input("fullname", id="field-fullname", label=_("Full Name"), placeholder=_("Joe Bloggs"), value=data.fullname, error=errors.fullname, classes=["control-medium"], attrs={'data-module': 'slug-preview-target', 'class': 'form-control'}, is_required=True) }}
+      {% set attrs = {'data-module': 'slug-preview-slug', 'data-module-placeholder': '<username>', 'class': 'form-control'} %}
+      {{ form.prepend('name', id='field-username', label=_('Username'), placeholder=_('username'), value=data.name, error=errors.name, attrs=attrs, is_required=true) }}
+    </div>
+    <style>
+      /* change URL: field to Username: */
+      #username-input-field .slug-preview strong {
+          display: none;
+      }
+      #username-input-field .slug-preview:before {
+          content: '{{ _("Username") }}:';
+          font-weight: bold;
+      }
+      /* make username input conform to form style */
+      #username-input-field .input-group{
+        width: 100%;
+      }
+      #username-input-field .form-control{
+        border-radius: 3px;
+      }
+    </style>
+  {% else %}
+    {{ form.input("name", id="field-username", label=_("Username"), placeholder=_("username"), value=data.name, error=errors.name, classes=["control-medium"], is_required=True) }}
+    {{ form.input("fullname", id="field-fullname", label=_("Full Name"), placeholder=_("Joe Bloggs"), value=data.fullname, error=errors.fullname, classes=["control-medium"]) }}
```

Comment:
add `is_required=true`
also add `not_empty` validator in `ckanext.emailasusername.plugin.emailasusername_new_user_schema` based on `ckan.logic.schema.default_user_schema``